### PR TITLE
Documentation branch, Index.py fixed

### PR DESCRIPTION
**Pull request by eugenetyc**
Hi guys,
Index.py now uses the (accumulated zonal TFs) for length calculation
Documentation is updated please review.

Also, remember to change your search.py parts!

Jiayou :D

### CODE REVIEW

**Review by hcwong — APPROVED** [review]
lgtm but some comments. but remember to change the usage of `.size` on line 302 and 303 or we will have bugs. Also remember to test before merging

File: index.py [hunk 1]
```diff
@@ -226,69 +248,84 @@ def generate_positional_indexes_from_list(self, words, start_index):
...
+                if (posting.doc_id) not in tf_overall:
+                    tf_overall[posting.doc_id] = tf_contribution
                 else:
-                    self.doc_lengths[posting.doc_id] += (posting_weight * posting_weight)
+                    tf_overall[posting.doc_id] += tf_contribution
+
+            # Since each term has a non-zero tf contribution to give a non-zero length contribution (due to lnc document weighting scheme)
+            # to the length of the document vector if the term appears in the document vector,
+            # we calculate this length contribution to the document length
+            for id, tf_value in D1.items():
```
Comment (hcwong) [inline index.py:286]: Cant seem to find D1's definituion, you mean tf_overall?

File: index.py [hunk 2]
```diff
@@ -226,69 +248,84 @@ def generate_positional_indexes_from_list(self, words, start_index):
...
+        # Square-root to find the vector length
         for doc_id, total_weight in self.doc_lengths.items():
             self.doc_lengths[doc_id] = math.sqrt(total_weight)
 
     def write(self):
         """
         Writes PostingList objects into postings file and all terms into dictionary file
-        Document lengths are also written into dictionary file
-        All doc_ids are also written into postings file
+        doc_lengths and docid_setare also written into dictionary file
```
Comment (hcwong) [inline index.py:300]: minor, typo

File: index.py [hunk 3]
```diff
@@ -304,28 +341,29 @@ def generate_string_of_posting(self):
...
-    A PostingList contains size and Postings, which have doc_id and freq of each term
+    Each PostingList is a collection of Postings for a particular term.
+    A PostingList contains the number of unique documents it contains regardless of which zone/field (size) and a list of Postings (postings)
     """
     def __init__(self):
         self.postings = []
-        self.size = 0  # number of insert / term frequency
+        self.unique_docids = 0
 
     def get_size(self):
```
Comment (hcwong) [inline index.py:351]: Should function name to get_unique_docids for clarity

File: index.py [hunk 4]
```diff
@@ -304,28 +341,29 @@ def generate_string_of_posting(self):
...
-        return self.size
+        return self.unique_docids
 
     # Insert with var byte encoding
-    def insert(self, doc_id, field, positions):
-        next_id = self.size
+    def insert(self, doc_id, field, positions, new_doc_id=True):
+        next_id = self.unique_docids
         new_posting = Posting(next_id, doc_id, field, positions)
```
Comment (hcwong) [inline index.py:357]: Note for future, remove index attribute from `Posting` class

File: index.py [hunk 5]
```diff
@@ -304,28 +341,29 @@ def generate_string_of_posting(self):
...
-        self.size += 1
+        if new_doc_id:
+            self.unique_docids += 1
 
     def insert_without_encoding(self, doc_id, field, positions):
-        next_id = self.size
+        next_id = self.unique_docids
         self.postings.append(Posting(next_id, doc_id, field, positions))
-        self.size += 1
+        self.unique_docids += 1
```
Comment (hcwong) [inline index.py:366]: We should do a check here, whether or not to increment `unique_docids`, like the one done in `insert` function

File: index.py (hunk 1 above)
Comment (eugenetyc) [reply index.py:286]: Ahh yes, oversight! Thanks!

**Pull request closed without merging**